feat: negotiate pixi build RPC interface through interface package

Hofer-Julian opened the pull request:
See also https://linear.app/prefix-dev/issue/PXI-13/inject-mutex-package-constraints-into-solve-of-build-backends-to

TODO:
- [ ] Document interface update procedure
- [ ] Tests

baszalmstra commented on crates/pixi_build_frontend/src/protocols/builders/pixi.rs:
```diff
@@ -195,17 +196,20 @@ impl ProtocolBuilder {
             .map_err(FinishError::SpecConversionError)?;
 
+        // Ensure that backend uses the same interface as Pixi
+        specs.push(PIXI_BUILD_INTERFACE_MATCHSPEC.clone());
```
Instead of spec we should add a constraint and then check if the package is part of the solution. Only then are we certain that a package in the solution depends on it and it was not pixi that introduced the package.

baszalmstra reviewed (commented):
Dont forget to add the implementation to the command dispatcher too!

baszalmstra commented on crates/pixi_build_frontend/src/tool/installer.rs:
```diff
@@ -232,14 +236,32 @@ impl ToolInstaller for ToolContext {
             .map(GenericVirtualPackage::from)
             .collect();
 
+        let build_api_version_spec = MatchSpec::from_str(
+            &format!("{PIXI_BUILD_API_NAME}=={PIXI_BUILD_API_VERSION}"),
+            ParseStrictness::Strict,
+        )
+        .expect("should always succeed");
```
I mean to have the lower bound and upper bound of the compatible APIs in `pixi_build_types` and the name of the interface package. You can construct a `NamelessMatchSpec` from the bounds and a matchspec from the `NamelessMatchSpec`.

```rust
let version_range = VersionSpec::Group(LogicalOperator::And, vec![
    VersionSpec::Range(RangeOperator::GreaterEquals, Version::major(lower_bound)),
    VersionSpec::Range(RangeOperator::Less, Version::major(upper_bound))
]);
let spec = NamelessMatchSpec {
    version: Some(version_range),
    .. NamelessMatchSpec::default()
};
let package_name = pixi_build_package_name;
let match_spec = MatchSpec::from_nameless(spec, Some(package_name));
```

Hofer-Julian commented on tests/integration_rust/command_dispatcher/mod.rs:
```diff
@@ -0,0 +1,102 @@
+        .await
+        .unwrap_err();
+
+    insta::assert_snapshot!(err);
```
What do you mean with `pixi_tools` :)

baszalmstra replied:
Ah my bad I meant `pixi_test_utils`. Specifically the function `format_diagnostic`.

Hofer-Julian commented on crates/pixi_command_dispatcher/src/instantiate_tool_env/mod.rs:
```diff
@@ -179,6 +187,22 @@ impl InstantiateToolEnvironmentSpec {
             .await
             .map_err(InstantiateToolEnvironmentError::UpdateLock)?;
 
+        let build_api_version_spec = MatchSpec::from_str(
+            &format!("{PIXI_BUILD_API_NAME}=={PIXI_BUILD_API_VERSION}"),
+            ParseStrictness::Strict,
+        )
+        .expect("should always succeed");
```
> Maybe we should rename pixi_build_types to pixi_build_api.

I think as-is the current name makes a bit more sense?
If it were called `pixi_build_api`, I would e.g. expect that `pixi_build_frontend::Protocol` would live in `pixi_build_api` instead.

Hofer-Julian commented:
Your comments should be addressed now.

As discussed:
- moved integration tests back to crate `pixi_command_dispatcher`
- dropped documentation task and tackle that in separate issue
